ipamd/ipamd.go

```diff
@@ -348,177 +352,254 @@ func (c *IPAMContext) StartNodeIPPoolManager() {
-	if err != nil {
-		log.Errorf("Failed to increase pool size due to not able to allocate ENI %v", err)
-
-		if isAttachmentLimitExceededError(err) {
-			c.maxENI = c.dataStore.GetENIs()
-			log.Infof("Discovered the instance max ENI allowed is: %d", c.maxENI)
+		eniInfos := c.dataStore.GetENIInfos()
+		eni := eniInfos.ENIIPPools[eniID]
+		err = c.awsClient.AllocIPAddresses(eniID, maxIPPerENI-int64(len(eni.IPv4Addresses)))
+		if err != nil {
+			log.Warnf("Failed to allocate all available ip addresses on an eni %s: %s", eni.ID, err)
+			// Continue to process the allocated IP addresses
+			ipamdErrInc("increaseIPPoolAllocAllIPAddressFailed", err)
 		}
-		// TODO need to add health stats
-		ipamdErrInc("increaseIPPoolAllocENI", err)
-		return
-	}
 
-	maxIPLimit, err := c.awsClient.GetENIipLimit()
-	if err != nil {
-		log.Infof("Failed to retrieve ENI IP limit: %v", err)
-		return
-	}
+		eniMetadata, err := c.waitENIAttached(eniID)
+		if err != nil {
+			ipamdErrInc("increaseIPPoolWaitENIAttachedFailed", err)
+			log.Errorf("Failed to increase pool size: not able to discover attached ENI from metadata service %v", err)
+			return
+		}
 
-	if warmIPTargetDefined {
-		err = c.awsClient.AllocIPAddresses(eni, int64(curIPTarget))
+		err = c.setupENI(eniID, eniMetadata)
+		if err != nil {
+			ipamdErrInc("increaseIPPoolSetupENIFailed", err)
+			log.Errorf("Failed to increase pool size: %v", err)
+			return
+		}
 	} else {
-		err = c.awsClient.AllocIPAddresses(eni, maxIPLimit)
-	}
-	if err != nil {
-		log.Warnf("Failed to allocate all available ip addresses on an ENI %v", err)
-		// Continue to process the allocated IP addresses
-		ipamdErrInc("increaseIPPoolAllocAllIPAddressFailed", err)
+		log.Debugf("Skipping ENI allocation due to max ENI already attached to the instance : %d", maxENIs)
 	}
 
-	eniMetadata, err := c.waitENIAttached(eni)
-	if err != nil {
-		ipamdErrInc("increaseIPPoolwaitENIAttachedFailed", err)
-		log.Errorf("Failed to increase pool size: not able to discover attached ENI from metadata service %v", err)
-		return
-	}
+	// For each ENI, try to fill in missing IPs
+	eniInfos := c.dataStore.GetENIInfos()
+	for _, eni := range eniInfos.ENIIPPools {
 
-	err = c.setupENI(eni, eniMetadata)
-	if err != nil {
-		ipamdErrInc("increaseIPPoolsetupENIFailed", err)
-		log.Errorf("Failed to increase pool size: %v", err)
-		return
+		maxIPPerENI, err := c.awsClient.GetENIipLimit()
+		if err != nil {
+			log.Infof("Failed to retrieve ENI IP limit: %v", err)
+			return
+		}
+
+		skipPrimary := UseCustomNetworkCfg()
+		if skipPrimary && eni.IsPrimary {
+			log.Debugf("Skip the primary ENI for need IP check")
+			continue
+		}
+		if int64(len(eni.IPv4Addresses)) < maxIPPerENI {
+			log.Debugf("Found ENI %s that has less than the maximum number of IP addresses allocated: cur=%d, max=%d",
+				eni.ID, len(eni.IPv4Addresses), maxIPPerENI)
+
+			err = c.awsClient.AllocIPAddresses(eni.ID, maxIPPerENI-int64(len(eni.IPv4Addresses)))
```

Comment:
Here is the IP allocation logic for each ENI.  We attempt to allocate the remaining IPs on the ENI, keep them around until the next `decreaseIPPool` (which is called if we are over WARM_IP_TARGET), and then they will be unassigned if they are still not used.  We only get here if we have less than `WARM_IP_TARGET`, so this doesn't continuously assign and unassign IPs.